---
title: "TCGAbrowser workflow"
author: "Phil Cheng"
date: "`r doc_date()`"
package: "`r pkg_ver('TCGAbrowser')`"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{TCGAbrowser workflow}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

# Introduction
The Cancer Genome Atlas has provided a comprehensive omics study of 33 cancers. The National Cancer Institute (NCI) has made this data pubically available on the Genomic Data Commons [(GDC)](https://gdc-portal.nci.nih.gov/) Data Portal. There are several R packages that download the data, `r Biocpkg("RTCGAToolbox")`, `r Biocpkg("RTCGA")`, `r Biocpkg("TCGAbiolinks")`, `r CRANpkg("TCGA2STAT")`, `r Biocpkg("MultiAssayExperiment")`, and [TCGA-Assembler](http://www.compgenome.org/TCGA-Assembler/).

The motivation for TCGAbrowser was to provide convenience functions and wrappers to analyze the omics datasets by gene expression or mutation subsetting. Typically, biologists are studying one gene in one cancer by overexpression and/or knockdown experiments. TCGAbrowser was designed to subset the datasets by gene expression or gene mutation and perform differential gene expresion, mutation, copy number, protein and survival analysis. 

# Installation
To install TCGAbrowser use the code below.
```{r, message = FALSE, warning = FALSE}
require(devtools)
devtools::install_github("pcheng84/TCGAbrowser")
require(TCGAbrowser)
```

The workflow for TCGAbrowser is shown in the graphic below.

```{r, out.width = "50%", echo = FALSE}
knitr::include_graphics("workflow.png")
```

#Data download and formating
This vignette assumes the user will use one of the R packages mentioned above to download the TCGA data. In this example, we used `r CRANpkg("TCGA2STAT")` to download the data.

```{r, message = FALSE, warning = FALSE}
require(TCGA2STAT)
#download RSEM normalized gene expression values and clinical data
rnaseq.skcm <- getTCGA(disease = "SKCM", data.type = "RNASeq2", clinical = TRUE)

#download binary matrix of mutations
mut.skcm <- getTCGA(disease = "SKCM", data.type = "Mutation", type = "somatic")

#download copy number values, remove Y chromosome genes
cnv.skcm <- getTCGA(disease = "SKCM", data.type = "CNV_SNP", filter = "Y")

#make data.table version of RNAseq data and shortens sample name
require(data.table)
rna <- data.table(Gene = rownames(rnaseq.skcm$dat), rnaseq.skcm$dat)

#renames RNAseq data
setnames(rna, sub("(TCGA-.*?-.*?-.*?)-.*", "\\1", colnames(rna)))

#reorders samples alphabetically
setcolorder(rna, order(colnames(rna)))

#make data.table version of mutation data
mut <- data.table(Gene = rownames(mut.skcm$dat), mut.skcm$dat)

#make data.table version of copy number data and sets thresholds for loss and gain at -0.2 and 0.2 respectively
cp <- data.table(Gene = rownames(cnv.skcm$dat), cnv.skcm$dat)
for (j in seq_len(ncol(cp))[-1])  {
    set(cp, which(cp[[j]] > 0.2),j,1)
    set(cp, which(cp[[j]] < -0.2),j,-1)
    set(cp, which(cp[[j]] >= -0.2 & cp[[j]] <= 0.2),j,0)
  }
  
setnames(cp, sub("(TCGA-.*?-.*?-.*?)-.*", "\\1", colnames(cp)))
setcolorder(cp, order(colnames(cp)))

#get rppa data

'%&%' <- function(a, b) paste(a, b, sep="")
getTCGA_rppa = function(disease){
  # disease = "BRCA"
  ldoc <- XML::htmlTreeParse("http://gdac.broadinstitute.org/runs/stddata__latest/", useInternalNodes = T)
  llinks = unlist(XML::xpathApply(ldoc, "//a[@href]", XML::xmlGetAttr, "href"))
  d.links = grep("/data/" %&% disease,llinks,value = T)
  ddoc = XML::htmlTreeParse(d.links, useInternalNodes = T)
  keyWord = "//a[contains(@href, 'RPPA')]"
  plinks = XML::xpathSApply(ddoc, keyWord, XML::xmlGetAttr, 'href')
  plinks = plinks[grepl("Level_3.*.gz$", plinks)]
  download.file(d.links %&% "/" %&% plinks, destfile = "temp.tar.gz")
  untar("temp.tar.gz")
  file.remove("temp.tar.gz")
  infile = sub("^(gdac.broadinstitute.org_.*).tar.gz$", "\\1", plinks) 
  #annot = fread(infile %&% "/" %&% disease %&% ".antibody_annotation.txt")
  if(file.exists(infile %&% "/" %&% disease %&% ".rppa.txt")){
    rppa = read.table(infile %&% "/" %&% disease %&% ".rppa.txt", sep="\t", head=T, stringsAsFactors = F)
    names(rppa) = gsub("\\.","-",names(rppa))
    t.ceref = t(sapply(rppa$'Composite-Element-REF', function(x) unlist(strsplit(x, "\\|")[[1]])))
    rppa$Gene = t.ceref[,1]
    rppa$CE.REF = t.ceref[,2]
    rppa = rppa[,c(ncol(rppa)-1,ncol(rppa),2:(ncol(rppa)-2))]
   } else {
    rppa = "file not found"
  } 
  unlink(infile, recursive=TRUE)
  rppa
}

rppa <- data.table(getTCGA_rppa("SKCM"))
setnames(rppa, sub("(TCGA-.*?-.*?-.*?)-.*", "\\1", colnames(rppa)))

#put survival dates into one column
pat <- data.table(bcr_patient_barcode = rownames(rnaseq.skcm$clinical), rnaseq.skcm$clinical)
pat$days <- as.numeric(ifelse(is.na(pat$daystodeath), pat$daystolastfollowup, pat$daystodeath)) 
pat[, TCGA_year := days/365.25]
pat[, vitalstatus := as.numeric(vitalstatus)]

#adds RNAseq sample name to clinical data
samp <- data.table(name = colnames(rna), bcr_patient_barcode = sub("(TCGA-.*?-.*?)-.*", "\\1", colnames(rna)))
pat <- merge(pat, samp, by = "bcr_patient_barcode")
```

## 
#Subset by gene expression
`rnasubset` needs the clinical table, the RNAseq count table, a gene and a percentage cutoff. It will return a data.table with 5 additional columns to the original clinical table. The level column will contain the RSEM values for the gene of interest. The exprs_rank column will contain the rank of gene expression. The high column will contain TRUE values for samples that are in the upper quantile of expression. The low column will contain TRUE values for samples that are in the lower quantile of expression. The gene2 column contains low, middle, high identifiers for which gene expression group the sample belongs to.


```{r}
sox10.rna <- rnasubset(pat, rna, "SOX10", 20)
```

#Expression plotting
The new table can be used for downstream analysis and plotting functions. `plotlygenelevel` and `hchartgenelevel` will use plotly and highcharter to plot the gene expression of your gene of interest across all samples with coloring for the low, middle and high groups. 

```{r}
plotlygenelevel(sox10.rna)
hchartgenelevel(sox10.rna)
```

#Survival Analysis
`genesurv` will plot a kaplan-meier plot using `r CRANpkg("survminer")` comparing the high and low group

```{r}
genesurv(sox10.rna, "SOX10")
```

#Differential gene expression analysis
`rnadeg` will use voom from the `r Biocpkg("limma")` package to calculate the genes differentially expressed between the high and low group

```{r}
sox10.deg <- rnadeg(sox10.rna, rna)
sox10.deg
```

`rnaheat` will plot a heatmap of the top 100 most significant differentially expressed genes with `r CRANpkg("ComplexHeatmap")`

```{r}
rnaheat(sox10.rna, rna, sox10.deg, "SOX10")
```

#GSVA pathway analysis
`rnagsva` will use `r Biocpkg("GSVA")` to calculate the GSVA enrichment scores for all samples. `rnagsvadeg` will use `r Biocpkg("limma")` to determine the significant differences in GSVA enrichment scores between the high and low group. `rnagsvaheat` will use `r CRANpkg("ComplexHeatmap")` to display a heatmap of the top 100 most signifcant pathways

```{r, message = FALSE, warning = FALSE}
require(GSVAdata)
data(c2BroadSets)
sox10.gsva <- rnagsva(sox10.rna, rna)
sox10.gsvasig <- rnagsvasig(sox10.rna, sox10.gsva)
rnagsvaheat(sox10.rna, sox10.gsva, sox10.gsvasig, "SOX10")
```

#Reactome pathway analysis
`rnareact` will use `r Biocpkg("ReactomePA")` to calculate enriched pathways from Reactome database. `plotreact` can plot a dotplot, enrichment pathway, or cnetmap of the significant reactome pathways.

```{r}
sox10.react <- rnareact(sox10.deg)

#dotplot
plotreact(sox10.react, sox10.deg, "dot", 15)

#pathway plot
plotreact(sox10.react, sox10.deg, "map", 15)

#cnet plot
plotreact(sox10.react, sox10.deg, "cnet", 5)
```

#Differential mutation analysis
`diffmut` will perform a chi-squared test between the proportions of mutated vs wild-type genes between the high and low groups. The binary mutation matrix is an over-simplification of mutation status of a gene. Here the assumption is every mutation in a gene causes the same effect. `plotlymut` will create an Oncoprint-like plot of the top 20 differentially mutated genes and your gene of interest.

```{r, warning = FALSE}
sox10.mut <- diffmut(sox10.rna, mut)
plotlymut(sox10.rna, mut, sox10.mut, "SOX10")
```

#Differential copy number analysis
`diffcp` will perform a chi-squared test between copy number gain and copy number neutral and another chi-squared test for copy number neutral and copy number loss. `plotlycp` will create an Oncoprint-like plot of the top 20 genes with copy number differenes and your gene of interest.

```{r, warning = FALSE}
sox10.cp <- diffcp(sox10.rna, cp)
plotlycp(sox10.rna, cp, sox10.cp, "SOX10")
```


#Differential RPPA analysis
`rppadeg` will match the samples from the high and low group and use `r Biocpkg("limma")` to determine differential protein expression. `rppaheat` will draw the heatmap for all significant differentially expressed proteins

```{r}
rppaheat(sox10.rna, rppa, "SOX10")
sox10.rppa <- rppadeg(sox10.rna, rppa)
```

