#' rnaheat function
#'
#' Uses ComplexHeatmap to draw heatmap
#'
#' @param pat2 data frame generated by patsubset function
#' @param rna data frame of RNAseq values
#' @param deg data frame of differential expressed genes from rnadeg function
#' @param gene character(1) Gene symbol of gene of interest
#'
#' @import data.table
#' @import ComplexHeatmap
#' @import circlize
#'
#' @return ComplexHeatmap of top 100 significant differentially expressed genes
#'
#' @examples
#' gene <- "SOX10"
#' sox10.pat <- patsubset(pat, rna, gene, 10)
#' sox10.deg <- rnadeg(sox10.pat, rna)
#' rnaheat(sox10.pat, rna, sox10.deg, gene)
#'
#' @export
#'
rnaheat <- function(pat2, rna, deg, gene) {
  setkey(pat2, gene2)
  rna.log <- data.table(Gene = rna$Gene, cpm(rna[,pat2[!("middle"), name], with=F], log=T, normalized.lib.sizes=F))
  setkey(rna.log, Gene)
  h1 <- as.matrix(rna.log[deg$genes[1:100], colnames(rna.log)[-1], with=F])
  rownames(h1) <- deg$genes[1:100]
  h1.t <- t(apply(h1, 1, scale))
  colnames(h1.t) <- colnames(h1)
  df <- data.frame(as.character(pat2[!("middle"),gene2]))
  colnames(df) <- gene
  col1 <- list(Cell = c("high" = "#ca0020",
                        "low" = "#0571b0"))
  names(col1) <- gene
  top_ha <- HeatmapAnnotation(df = df, col = col1)

  Heatmap(h1.t, top_annotation = top_ha, name = "color scale", show_column_names = T, row_names_gp = gpar(fontsize=8), column_names_gp = gpar(fontsize=6))
 # for(an in colnames(df)) {
  #  decorate_annotation(an, {
      # annotation names on the right
   #   grid.text(an, unit(1, "npc") + unit(2, "mm"), 0.5, default.units = "npc", just = "left")
      # annotation names on the left
      #grid.text(an, unit(0, "npc") - unit(2, "mm"), 0.5, default.units = "npc", just = "right")
    #})
  #}
}
